server/src/main/java/org/jboss/as/server/deploymentoverlay/AffectedDeploymentOverlay.java

```diff
@@ -0,0 +1,212 @@
+        Set<String> links = new HashSet<>();
+        for(String serverGoupName : serverGoupNames) {
+            links.addAll(listLinks(context, PathAddress.pathAddress(
+                    PathElement.pathElement(SERVER_GROUP, serverGoupName),
+                    PathElement.pathElement(DEPLOYMENT_OVERLAY, overlay))));
+        }
+        return links;
+    }
+
+    public static Set<String> listLinks(OperationContext context, PathAddress deploymentsRootAddress) {
+        Resource overlayResource = context.readResourceFromRoot(deploymentsRootAddress);
+        if (overlayResource.hasChildren(DEPLOYMENT)) {
+            return overlayResource.getChildrenNames(DEPLOYMENT);
+        }
+        return Collections.emptySet();
+    }
+
+    public static void redeployLinks(OperationContext context, PathAddress deploymentsRootAddress, Set<String> runtimeNames) throws OperationFailedException {
+        Set<String> deploymentNames = listDeploymentNames(context, deploymentsRootAddress, runtimeNames.stream().map(wildcardExpr -> Pattern.compile(wildcardToJavaRegex(wildcardExpr))).collect(Collectors.toSet()));
+        for (String deploymentName : deploymentNames) {
+            PathAddress address = deploymentsRootAddress.append(DEPLOYMENT, deploymentName);
+            OperationStepHandler handler = context.getRootResourceRegistration().getOperationHandler(address, REDEPLOY);
+            ModelNode operation = addRedeployStep(address);
+            assert handler != null;
+            assert operation.isDefined();
+            context.addStep(operation, handler, OperationContext.Stage.MODEL);
+        }
+    }
+
+    public static void redeployLinksAndTransformOperationForDomain(OperationContext context, Set<String> serverGroupNames, Set<String> runtimeNames) throws OperationFailedException {
+        Set<String> deploymentNames = new HashSet<>();
+        for (String serverGoupName : serverGroupNames) {
+            deploymentNames.addAll(listDeploymentNames(context, PathAddress.pathAddress(PathElement.pathElement(SERVER_GROUP, serverGoupName)), runtimeNames.stream().map(wildcardExpr -> Pattern.compile(wildcardToJavaRegex(wildcardExpr))).collect(Collectors.toSet())));
+        }
+        //Remote part
+        Operations.CompositeOperationBuilder opBuilder = Operations.CompositeOperationBuilder.create();
+        deploymentNames.forEach((deploymentName) -> opBuilder.addStep(addRedeployStep(context.getCurrentAddress().getParent().append(DEPLOYMENT, deploymentName))));
+        // Add the domain op transformer
+        List<DomainOperationTransformer> transformers = context.getAttachment(OperationAttachments.SLAVE_SERVER_OPERATION_TRANSFORMERS);
+        if (transformers == null) {
+            context.attach(OperationAttachments.SLAVE_SERVER_OPERATION_TRANSFORMERS, transformers = new ArrayList<>());
+        }
+        final ModelNode slave = opBuilder.build().getOperation();
+        transformers.add(new OverlayOperationTransformer(slave));
+    }
```

Comment:
Unfortunately, DomainOperationTransformer is an extremely poorly named thing used for a quite specialized thing -- replacing content params without a "hash" field with ones with a "hash" field. It wasn't original meant as a "transformer" in the sense that's needed here, i.e. something that can translate an op that a legacy slave in a mixed domain doesn't understand into one it does understand, or produce a user friendly error message if that's not possible.

It looks like you've repurposed DomainOperationTransformer to reuse it for mixed domain. I suspect it will work too! But it makes me really nervous adding a whole new pattern to how we do these things. People won't remember it and are therefore likely to break it.